File: pal/model/x86_64/access_mechanism/cpuid.py
from pal.model.access_mechanism import AbstractAccessMechanism
from dataclasses import dataclass

@dataclass()
class CPUID(AbstractAccessMechanism):
    """ Access mechanism for reading information from the CPUID instruction """

    leaf: bytes = 0
    """ The CPUID leaf to be read (eax)  """

    output: str = "eax"
    """ The output register that CPUID reads information into  """
    """ Options: eax, ebx, ecx, edx  """

    name: str = "cpuid"
    """ The name of this access mechanism  """

    def is_read(self):
        return True

    def is_write(self):
        return False

    def is_valid(self):
        if str(output).lower() not in ["eax", "ebx", "ecx", "edx"]:
            return False

        # TODO: Check leaf/subleaf within valid range

        return True

File: test/c/intel_x64/cpuid.c
#include <pal/instruction/cpuid.h>
#include <pal/cpuid/leaf_01_eax.h>

void test_cpuid_compile(void)
{
    uint32_t leaf = 0x1;
    uint32_t subleaf = 0x0;

    pal_cpuid_return_values output = pal_execute_cpuid(leaf, subleaf);

    uint32_t eax = output.eax;
    uint32_t ebx = output.ebx;
    uint32_t ecx = output.ecx;
    uint32_t edx = output.edx;

    pal_print_leaf_01_eax_from_value(eax);
    uint32_t family_id = pal_get_leaf_01_eax_family_id_from_value(eax);
}
